src/IB/IBFESurfaceMethod.cpp

```diff
@@ -640,48 +779,389 @@ IBFESurfaceMethod::interpolateVelocity(const int u_data_idx,
+                {
+                    for (unsigned int d = 0; d < NDIM; ++d)
+                        get_values_for_interpolation(
+                            DU_j_node[d], *DU_j_petsc_vec[d], DU_j_local_soln[d], DU_j_dof_indices[d]);
+                    // get_values_for_interpolation(DU_j_node[d], *DU_j_ghost_vec[d], DU_j_dof_indices[d]);
+                }
+
                 const bool qrule_changed =
                     FEDataManager::updateInterpQuadratureRule(qrule, d_default_interp_spec, elem, x_node, patch_dx_min);
                 if (qrule_changed) fe->attach_quadrature_rule(qrule.get());
+
                 fe->reinit(elem);
                 const unsigned int n_node = elem->n_nodes();
                 const unsigned int n_qp = qrule->n_points();
+
+                // Zero out the values prior to accumulation.
                 double* x_begin = &x_qp[NDIM * qp_offset];
                 std::fill(x_begin, x_begin + NDIM * n_qp, 0.0);
-                for (unsigned int k = 0; k < n_node; ++k)
+                boost::array<double*, NDIM> DU_j_begin;
+                for (unsigned int d = 0; d < NDIM; ++d)
                 {
-                    for (unsigned int qp = 0; qp < n_qp; ++qp)
+                    DU_j_begin[d] = &DU_j_qp[d][NDIM * qp_offset];
+                    std::fill(DU_j_begin[d], DU_j_begin[d] + NDIM * n_qp, 0.0);
+                }
+
+                double* N_begin = &N_qp[NDIM * qp_offset];
+                std::fill(N_begin, N_begin + NDIM * n_qp, 0.0);
+
+                // Interpolate X, du, and dv at all of the quadrature points
+                // via accumulation, i.e., X(qp) = sum_k X_k * phi_k(qp) for
+                // each qp.
+
+                //~
+                for (unsigned int qp = 0; qp < n_qp; ++qp)
+                {
+                    for (unsigned int k = 0; k < NDIM - 1; ++k)
                     {
-                        const double& p = phi[k][qp];
-                        for (unsigned int d = 0; d < NDIM; ++d)
+                        interpolate(dX_dxi[k], qp, X_node, *dphi_dxi[k]);
+                        interpolate(dx_dxi[k], qp, x_node, *dphi_dxi[k]);
+                    }
+                    if (NDIM == 2)
+                    {
+                        dX_dxi[1] = VectorValue<double>(0.0, 0.0, 1.0);
+                        dx_dxi[1] = VectorValue<double>(0.0, 0.0, 1.0);
+                    }
+                    N = (dX_dxi[0].cross(dX_dxi[1])).unit();
+                    n = (dx_dxi[0].cross(dx_dxi[1])).unit();
+
+                    for (unsigned int i = 0; i < NDIM; ++i)
+                    {
+                        for (unsigned int k = 0; k < n_node; ++k)
                         {
-                            x_qp[NDIM * (qp_offset + qp) + d] += x_node[k][d] * p;
+                            x_qp[NDIM * (qp_offset + qp) + i] += x_node[k][i] * phi[k][qp];
                         }
+                        N_qp[NDIM * (qp_offset + qp) + i] = n(i);
```

Comment:
Please don't mix uppercase N and lower case n --- they have different meanings. (N indicates reference coordinates, n indicates current coordinates.)

Reply:
Okay I see what you mean. I'll fix this